### DISCUSSION TITLE
Automatic moderation functionality - wish list?

### DISCUSSION THREAD

**Discussion by vadim-berman** [EssentialsX/Essentials#5777] [Ideas and Feature Suggestions] [answered]
Hello EssentialX community!

I am the CEO of Tisane Labs, a content moderation API vendor. After discussions with several MC hosting companies we felt that there is an unfulfilled need for moderation that is beyond keyword spotting. But instead of building yet another plugin, we think that piggybacking on EssentialX would make it easier to deploy and adopt. 

We want to add a way for EssentialX to connect seamlessly to Tisane. 

This is what we can detect:

* obfuscated profanities
* racism (even if not using slurs)
* sexual advances
* attempts to sell drugs
* ad hominems / insults
* suicidal ideation
* topics and named entities (e.g. if you don't want the users to talk about politics and start fights)

We also support so-called [2-factor moderation](https://tisane.ai/knowledgebase/2-factor-moderation/) which takes the burden of resolving petty conflicts. 

You can pick and choose the how to handle each type of violations. 

(Goes without saying, no one is perfectly correct all the time but we also provide tools to track and handle errors.)

The question is, what would you like to be added to EssentialX moderation-wise? Any sort of wish list features.

**Comment by codebycam** [comment]
Not affiliated with EssentialsX at all, however I have been using their software for over 10 years. This sounds like a horrible idea, and just seems like you're trying to push your [very expensive](https://tisane.ai/subscription-plans/) product.

**Reply by vadim-berman** [reply]
Yes, seamless integration is promotion (shocking, no?).
No, it's not even remotely "expensive", it's solidly in the budget tier.

**Reply by codebycam** [reply]
$100(USD?) a month as the cheapest tier (which isn't free) isn't expensive? Your business must be booming, mate! As a small server owner (which a large portion of the servers using EssentialsX would be), $100 is a _lot_ per month. 

Good luck in your quest!

**Reply by vadim-berman** [reply]
The cheapest tier is, as you may have noticed, free. As in, **forever free**. Try as hard as I can, I can't lower it below 0. Yes, it's rate limited, but for low traffic, it works. 

If your traffic is low but you don't want to be rate-limited, then you can go to [API Layer where it's per piece](https://apilayer.com/marketplace/tisane-api).

US$100 for 100,000 requests per month is ridiculously low for a commercial moderation API. Most of our competitors start at US$3,000, and even that is hard to maintain. 

The business is growing, yes. I am not entirely sure how a "mate" can survive with a business that finds $100 a month expensive.

**Reply by smashyalts** [reply]
a majority of mc servers arent a business, 100 usd a month, can give you a full dedicated machine, one of the best antiddos there is for mc's next highest plan, 100 usd is a heck of a lot.

**Reply by smashyalts** [reply]
a majority of servers operations costs are a lot lower than 100 usd a month

**Reply by vadim-berman** [reply]
I was replying to the poster above who says he runs a small business, appears to be from Australia, and astonishingly claims $100 is a lot. 

Appreciate the info. What we're aiming at is large-scale hosting, who may then either pass on the costs as an optional add-on or waive it, whatever they like. If you host, say, 50 servers and charge US$2 each for premium capabilities, it's affordable enough for Sierra Leone.

**Reply by smashyalts** [reply]
he... never said business?

**Reply by smashyalts** [reply]
he was talking about YOUR business

**Reply by vadim-berman** [reply]
Let's try again.  

>**As a small server owner (which a large portion of the servers using EssentialsX would be)**

**Reply by smashyalts** [reply]
a majority of minecraft servers ARE NOT A BUSINESS.

**Reply by codebycam** [reply]
His last comments just go to show this person has no clue about the Minecraft server scene, and hasn't bothered to put in any effort observing and learning it before trying to make some wild connection haha.

**Reply by underscore11code** [reply]
Let's give some data to back up Camm's claims of budget. As per the [EssentialsX Bstats](https://bstats.org/plugin/bukkit/EssentialsX/858), at the time of writing, the 1 month peak of players / servers is 94,084 / 64,112, which works out to an average of 1.46 players per server. However, this naive average assumes all servers are equal, which they most certainly are not. Most of those players will be clustered on a handful of larger servers, meaning the majority (no data to say exactly how much, but significantly more than half) will be running an average of 0-1 concurrent players. These kinds of servers are typically either free online hosting i.e. aternos, minehut, and friends; cheap shared hosting (to the tune of $1-$10/mo); or just running on a computer at home. For these kinds of servers, which the vast majority of the userbase are using, $100/mo is an absolutely insane price point that is far, FAR out of reach of what's sustainable.

edit: 6 month peak -> 1 month peak, misread bstats.

**Reply by vadim-berman** [reply]
@CodeByCam: have you heard the one about playing certain type of games and winning certain types of prizes?

**Reply by codebycam** [reply]
Shhhh shh shh sh, there there, @vadim-berman, you've embarrassed yourself enough.
![tenor](https://github.com/EssentialsX/Essentials/assets/41496198/aa552ba1-a48d-4586-9b34-708b5231b71b)

**Reply by vadim-berman** [reply]
@underscore11code: big thanks for the number crunching. 

Obviously, the distribution is always uneven, that is fine and expected. Very much like on a plane, the couple of guys in the business class pay for the bulk of the flight. So maybe 85% of the servers can either use the free plan (or even have communities too small for it to be usable), and 15% are commercial. 

It's a bit odd though. Either the folks we were talking about are heavily exaggerating or the stats are out of whack; or maybe EssentialX is not deployed as widely, in which case indeed it may not be a good idea. We were told though that the volume of conversations is enormous and they were aiming at the top level offering. That was the reason we're looking at the adventure in the first place. Given how their companies are growing though, I am inclined to believe them. 

I wonder if the maintainers of EssentialX would be happy to get their community growing thanks to a commercial plugin with a free option. I hope, the mindset here is not "cutting off the nose to spite one's face".

**Comment by electronicboy** [comment]
Was going to say, I'd much prefer that free, open-source projects within our community wasn't use to railroad somebody else's commercial software. 

There are already 1 or 2 options out there offering full plugins and platforms if people want them, compete on the fair market instead of trying to hijack your way into the ecosystem.

**Reply by vadim-berman** [reply]
You don't have to use it, if you don't like it. How is that even remotely "hijacking"? People with ginormous traffic need this option, otherwise we wouldn't bother. 

Can you elaborate on the "full plugins" please? Of course, we can build a separate one, but it's handier for the users not to install a different one. This is how ecosystems work. 

I'm fine with hostility, just scratching my head at the bizarre reasoning.

**Reply by pop4959** [reply]
I'm not sure where you see hostility in any of these messages, but rightfully so your post comes across as extremely out of touch with the open source community, and an addition that is clearly and obviously unwelcome in a plugin such as this one.

To me at least, it would appear that you have no clue what this plugin even does, and just decided to reach out with a business opportunity solely based off of the existing community that exists here.

**Reply by vadim-berman** [reply]
We are new to this, yes, hence the question. 

I decided to explore a business opportunity solely based off:

* my conversation with hosting people who need this functionality
* straightforward logic that installing one plugin is easier than installing two plugins
* knowledge that this plugin has moderation functionality

Assume you're in my place. What would be the right course of action, in your opinion?

1. Build a new plugin and make the admins install both plugins, reducing usability because... what exactly... something something out of touch with the open source spirit?
2. Seamlessly integrate both, providing a new feature that some people may use for free and others may pay for premium services.

**Reply by pop4959** [reply]
If I'm in your place I'd frankly choose option 1 as something more based in reality. Option 2 is completely out of the question given the circumstances, and only makes sense when you have no commitment to the community and are just looking to make a quick buck.

Another word of advice is to not insult one of the project maintainers when trying to close a deal.

**Reply by vadim-berman** [reply]
Appreciate the advice, but, FYI - in an admin community the reactions are the exact opposite. 

I am not trying to "close a deal" here, but to get feedback. Also, the proverbial buck is never quick and, as noted above there is also a free plan. 

If I may ask. If everyone here doesn't like commercial solutions linked to EssentialX, why did you choose a license that enables it?

**Reply by pop4959** [reply]
This is an admin community. The GPL is also a copyleft license - hopefully this provides clarity as to why we are required to use it.

**Reply by vadim-berman** [reply]
Sorry, bad wording. 

A community of MC admins, not related to plugins and such. 

I see. So it's literally just someone likes/doesn't like kind of things. Well, it is what it is.

**Comment by lordofpipes** [comment]
It would be better to approach this with a proposal of what a generalized chat moderation integration might require on the side of Essentials. I figure that anything that is overly specific needs to be justified with examples for how other alternative integration might be done, since otherwise it would be linking Essentials too closely to a commercial project most users are not interested in.

**Reply by vadim-berman** [reply]
Thanks. 

That's pretty much what the question is about, hence I'm asking about the wish-list. 

>Anything that is overly specific needs to be justified with examples for how other alternative integration might be done, since otherwise it would be linking Essentials too closely to a commercial project most users are not interested in.

I don't know if the question was misread or poorly formulated, but the idea is that **we** build the plugin. We're not asking anyone else to build it: 

* We build it.
* We distribute it.
* We publish it in a separate repository; whoever wants, can copy the integration, maybe tweak it for their own needs, for Tisane or any other API they want to use.

The question is, what's missing? What are the headaches?

**Comment by pop4959** [comment]
Locking things for now since things are getting too heated here.

**Comment by mdcfe** [comment, marked as answer]
It's unclear which of the following you're asking (and it seems you've asked for all three at different times):
* For the EssentialsX team to develop an integration with your service
* For the EssentialsX team to consider accepting a PR to integrate your service 
* For advice on whether you should fork EssentialsX to integrate your service

EssentialsX is a project intended for small-scale servers, almost all of which are run at minimal cost and are non-commercial. It provides very simple moderation features for staff to mute, kick and ban players on their server, as well as filtering out URLs and allowing users to block one another. There may be scope for additional chat moderation features, but these would still need to be accessible primarily to the above kinds of server owners (for example, more customisable text filters using regular expressions).

It is highly unlikely we would implement an integration for a third-party commercial moderation service within EssentialsX, as it serves little purpose for most users. Even in the event we were to integrate a third-party commercial service for the smaller segment of EssentialsX users who are commercial server owners, we would likely support an established service such as Community Sift, which is already trusted by several commercial server owners (and more widely online games as a whole), rather than a relatively unknown service/operator. There is similarly little interest in us accepting a PR to support this, as that code would become our responsibility to maintain despite being used by a select handful of commercial server owners.

There is no real reason why you would need to ship an integration with your service within EssentialsX. Processing of chat messages can easily be achieved in a standalone plugin using the Spigot and/or Paper APIs, and you can optionally call the EssentialsX API if you want to interact with any of its features (like private messages, mails, nicknames, mutes etc).

There isn't really any scope for EssentialsX to be involved here. Your subsequent responses make this pretty clear - it seems this discussiom mostly serves to promote a commercial service you are providing. I am sure your team is capable of researching the wider Minecraft server community and developing an appropriate integration for your service, but this isn't the place for that.
